### ISSUE TITLE
type struct parsing enhancements

### ISSUE THREAD

**Issue by stonedreamforest** [x64dbg/x64dbg#1305]
1.
a. if `.h ` have `#pragma once ` the "struct"  error:`line 1, col 1: unexpected character '#'`

b. if `.h ` have `array[0x50-0x30](or array[0x50*0x30]/array[0x50+0x30]/array[(0x50*0x30)/4])` the "struct"  error:`[line 78:5] expected ], got -` or `expected number token` or other errors

c. if `.h ` have 
```
typedef int( __fastcall *ExampleFunction )(char *arg1 , int arg2);  //or __stdcall or __thiscall
struct ExampleStruct {
	ExampleFunction			Fun;		//...
};
```
error [line 1:7] token ( is not a type...

d. if i` alt+f2` the “struct windows” not clear

r. if i chang` .h `struct  the “struct windows”  not refresh.    i already use` f5 `and load .h again

f. if i had parse` .h `the x64dbg did not remember it

g. the "struct wiodnws" `value option` can't display hex（i hope can Select a particular item display `dec` or `hex`）

h. the "struct wiodnws" have not ` real-time update `

i.  var type can't surport __int64(or __int32 \__int16\PVOID\UINT32\QWORD  )and more...

**suggestion:** surport include other header file.  for example：
```
//sys header
#include "windows.h"
#include "windns.h"
#include <winsock.h>
#include <emmintrin.h>
#include "xmmintrin.h"
#include <math.h>

//user header
#include "ExampleHeader1.H"		
#include "ExampleHeader2.H"		
#include "ExampleHeader3.H"	
```

2.
if first attach exe(or breakpoint) the `stack windows ` not Refresh

3.
such as address “0x000000006463218**0**” 
 i set  first memory breakpoint  suncess 
but i set secnond  memory breakpoint at   address “0x000000006463218**8**”    failed 
error ：Memory breakpoint already set!

**Comment by mrexodia** [comment]
1a. There are no plans to support `#pragma` statements
1b. There are no plans to support constant expression optimizations or any kind of expressions.
1c. I will eventually add support for function definitions, for now use `void*`
1d. That is a good point, it will be adjusted
1r. The refresh button will not reload any kind of files from disk (in fact the original location is not remembered) it is only used to refresh the values displayed.
1f. For now you can add a global initialization script that calls the `ParseTypes` command for all scripts you want to load on start.
1g. This is planned, currently I chose the most sensible defaults but probably I'll change them to display multiple types and eventually you can choose your own display type.
1h. For now this is avoided because it might take a lot of processing power for no reason, you can use F5 to manually refresh.
1i. You can use `typedef int64_t __int64;` to define your own types. [Here](https://github.com/x64dbg/x64dbg/blob/development/src/dbg/types.cpp#L20) you can see the primitive types supported or you can use `EnumTypes`
1j. You have to use a C preprocessor. See [here](https://msdn.microsoft.com/en-us/library/8z9z0bx6.aspx) or [here](http://stackoverflow.com/questions/3742822/preprocessor-output) for more information. I have no plans to write a preprocessor myself.
2. I opened a separate issue, see #1306 
3. This is not a bug you can place memory breakpoints on (a multiple of) 0x1000 pages only.

**Comment by stonedreamforest** [comment]
1r:
if .h have "struct"  for Example:
```
struct ExampleStruct{
	bool a;
};
```
now, i add new var:
```
struct ExampleStruct
{
	bool a;
	int	b;		//new
};
```
or
```
struct ExampleStruct
{
	int a;	//chang here (chang type or var name)
};
```
reload .harderr	->parse header	->visit type->.....

the struct windows not chang(add var or chang type or chang name) 
you can try it

**Comment by mrexodia** [comment]
The struct window is not meant to change and I believe you already
mentioned this in your first post...

**Comment by mrexodia** [comment]
However, I will check it out...

**Comment by stonedreamforest** [comment]
char often display error  if too bigger or to small
```
struct ExampleStruct
{
	char	ExampleChar[10];
};
```
**struct windows**:
```
			NAME						ADDR				SIZE		VALUE
   char ExampleChar[10]			00000000002af9f8			0XA			"abcdefg"
		ExampleChar[0]													97 'a'		// 97(dec) can dispaly hex
		ExampleChar[1]													98 'b'	
		ExampleChar[2]													99 'c'	
		ExampleChar[3]													100 'd'	
		ExampleChar[4]													101 'e'	
		ExampleChar[5]													102 'f'	
		ExampleChar[6]													103 'g'	
		ExampleChar[7]													0 '\0'	
		ExampleChar[8]													0 '\0'	
		ExampleChar[9]													0 '\0'	
```

**Comment by mrexodia** [comment]
some improvements have been made. visiting types should also be much much faster now...

**Comment by mrexodia** [comment]
Also, thanks for your feedback!

**Comment by stonedreamforest** [comment]
Faster than before 
but i test use：
```
struct ExampleStruct
{
	int8_t array[0x10000];
};
```
still very slow....

**Comment by mrexodia** [comment]
Yes it won't become much faster because it has to allocate 0x10000 nodes
which takes time.

**Comment by stonedreamforest** [comment]
and this：
![1](https://cloud.githubusercontent.com/assets/16742566/20793496/3efabe10-b802-11e6-91c4-2356f368c1f8.png)

```
{
	UnionTest u;	//address: 00007FFC23435574
	UnionTest u1;	//address: 00007FFC2343557c	+0x8
	UnionTest u2;	//address: 00007FFC23435574	+0x8
}
```

**Comment by mrexodia** [comment]
Why are you closing this?

**Comment by u07** [comment]
Is it possible to add enum support? Like this:
